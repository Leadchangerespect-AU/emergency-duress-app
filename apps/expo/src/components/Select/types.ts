import { ViewProps } from "react-native";

/** A single option in the Select. */
export interface SelectOption {
  label: string;
  value: string;
}

/** Props for the Select component. */
export interface SelectProps extends Omit<ViewProps, "children"> {
  /** List of selectable options (label/value pairs). */
  options: SelectOption[];
  /** Currently selected value, or null if none is selected yet. */
  value: string | null;
  /** Callback when the user selects a new value. */
  onChange: (newValue: string) => void;
  /**
   * Called when the select gains focus (e.g. user clicks/focuses on it on web,
   * or focuses the Pressable on iOS/Android).
   */
  onFocus?: () => void;
}
