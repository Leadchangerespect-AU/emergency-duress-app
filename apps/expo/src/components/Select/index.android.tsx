import React, { useState } from "react";
import { Modal, Pressable, Text, View } from "react-native";

import { SelectProps } from "./types";

/**
 * Android-specific Select:
 * Opens a Modal with a list of options when pressed.
 */
function Select({ options, value, onChange, onFocus }: SelectProps) {
  const [modalVisible, setModalVisible] = useState(false);

  const selectedOption = options.find((o) => o.value === value);
  const label = selectedOption ? selectedOption.label : "Select...";

  const openModal = () => setModalVisible(true);
  const closeModal = () => setModalVisible(false);

  const handleOptionPress = (optionValue: string) => {
    onChange(optionValue);
    closeModal();
  };

  return (
    <>
      <Pressable
        className="rounded border bg-white px-4 py-2"
        onPress={openModal}
        onFocus={onFocus}
      >
        <Text className="text-black">{label}</Text>
      </Pressable>

      <Modal visible={modalVisible} transparent onRequestClose={closeModal}>
        <View className="flex-1 items-center justify-center bg-black/50">
          <View className="w-3/4 rounded bg-white">
            {options.map((opt) => (
              <Pressable
                key={String(opt.value)}
                className="border-b border-gray-200 p-4"
                onPress={() => handleOptionPress(opt.value)}
              >
                <Text>{opt.label}</Text>
              </Pressable>
            ))}
            <Pressable onPress={closeModal} className="bg-gray-100 p-4">
              <Text className="text-center font-semibold">Cancel</Text>
            </Pressable>
          </View>
        </View>
      </Modal>
    </>
  );
}

export default Select;
