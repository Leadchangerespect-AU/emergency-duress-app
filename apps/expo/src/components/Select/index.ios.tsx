import React from "react";
import { ActionSheetIOS, Pressable, Text } from "react-native";

import { SelectProps } from "./types";

/**
 * iOS-specific Select:
 * Uses ActionSheetIOS to display a native action sheet when pressed.
 */
function Select({ options, value, onChange, onFocus, ...props }: SelectProps) {
  const selectedOption = options.find((o) => o.value === value);
  const label = selectedOption ? selectedOption.label : "Select...";

  const handlePress = () => {
    const optionLabels = options.map((o) => o.label);
    // Add "Cancel" at the end
    optionLabels.push("Cancel");
    const cancelButtonIndex = optionLabels.length - 1;

    ActionSheetIOS.showActionSheetWithOptions(
      {
        options: optionLabels,
        cancelButtonIndex,
      },
      (index) => {
        if (index === cancelButtonIndex) return;
        onChange(options[index].value);
      },
    );
  };

  return (
    <Pressable
      onPress={handlePress}
      onFocus={onFocus}
      className="rounded border bg-white px-4 py-2"
      aria-labelledby={props["aria-labelledby"]}
      accessibilityLabel={props.accessibilityLabel}
      id={props.id}
      testID={props.testID}
    >
      <Text className="text-black">{label}</Text>
    </Pressable>
  );
}

export default Select;
