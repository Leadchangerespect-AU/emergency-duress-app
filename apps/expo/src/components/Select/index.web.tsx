import React from "react";
import { Entypo } from "@expo/vector-icons";

import { SelectProps } from "./types";

/**
 * A native <select> element for the web platform.
 */
function Select({ options, value, onChange, onFocus, ...props }: SelectProps) {
  return (
    <div className="relative w-full">
      <select
        className="w-full appearance-none rounded-md bg-white py-2.5 pl-3 pr-8 text-base text-gray-900 outline outline-1 -outline-offset-1 outline-gray-300 focus:outline focus:outline-2 focus:-outline-offset-2 focus:outline-indigo-600 sm:text-sm/6"
        value={value ?? ""}
        onChange={(e) => {
          onChange(e.target.value);
        }}
        onFocus={onFocus}
        aria-labelledby={props["aria-labelledby"]}
        role="combobox"
      >
        {/* Optional placeholder if no value is selected */}
        {value === null || value === "" ? (
          <option value="" disabled>
            Select...
          </option>
        ) : null}
        {options.map((opt) => (
          <option key={opt.value} value={opt.value}>
            {opt.label}
          </option>
        ))}
      </select>
      <Entypo
        name="chevron-thin-down"
        className="pointer-events-none absolute right-2 top-1/2 -translate-y-1/2 transform text-gray-500 sm:size-4"
      />
    </div>
  );
}

export default Select;
