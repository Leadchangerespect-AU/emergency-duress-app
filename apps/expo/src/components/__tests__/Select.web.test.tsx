import React from "react";
import { render, screen } from "@testing-library/react-native";

import Select from "../Select/index.web";
import type { SelectProps } from "../Select/types";

jest.mock("@expo/vector-icons", () => ({
  Entypo: "",
}));

describe("Select (Web)", () => {
  const defaultProps: SelectProps = {
    options: [
      { label: "Option 1", value: "1" },
      { label: "Option 2", value: "2" },
      { label: "Option 3", value: "3" },
    ],
    value: null,
    onChange: jest.fn(),
  };

  // TODO add more tests

  it("matches snapshot", () => {
    render(<Select {...defaultProps} />);
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
