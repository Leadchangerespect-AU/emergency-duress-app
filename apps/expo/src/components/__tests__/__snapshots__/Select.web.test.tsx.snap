// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select (Web) matches snapshot 1`] = `
<div
  className="relative w-full"
>
  <select
    className="w-full appearance-none rounded-md bg-white py-2.5 pl-3 pr-8 text-base text-gray-900 outline outline-1 -outline-offset-1 outline-gray-300 focus:outline focus:outline-2 focus:-outline-offset-2 focus:outline-indigo-600 sm:text-sm/6"
    onChange={[Function]}
    role="combobox"
    value=""
  >
    <option
      disabled={true}
      value=""
    >
      Select...
    </option>
    <option
      value="1"
    >
      Option 1
    </option>
    <option
      value="2"
    >
      Option 2
    </option>
    <option
      value="3"
    >
      Option 3
    </option>
  </select>
  <
    className="pointer-events-none absolute right-2 top-1/2 -translate-y-1/2 transform text-gray-500 sm:size-4"
    name="chevron-thin-down"
  />
</div>
`;
